### PR TITLE
[dv] Use array for multi-reg in RAL model

### PR DESCRIPTION
**Pull request by weicaiyang**
Address #5852 item 1

Change to generate multi-reg as below. User can access the reg by using
the index, e.g. ral.sw_binding[0]
```
 class keymgr_reg_block extends dv_base_reg_block;
    rand keymgr_reg_sw_binding sw_binding[8];
    rand keymgr_reg_salt salt[8];
```

before this change, it generates the multi-reg like this
```
 class keymgr_reg_block extends dv_base_reg_block;
    rand keymgr_reg_sw_binding_0 sw_binding_0;
    rand keymgr_reg_sw_binding_1 sw_binding_1;
    rand keymgr_reg_sw_binding_2 sw_binding_2;
    rand keymgr_reg_sw_binding_3 sw_binding_3;
    rand keymgr_reg_sw_binding_4 sw_binding_4;
    rand keymgr_reg_sw_binding_5 sw_binding_5;
    rand keymgr_reg_sw_binding_6 sw_binding_6;
    rand keymgr_reg_sw_binding_7 sw_binding_7;
    rand keymgr_reg_salt_0 salt_0;
    rand keymgr_reg_salt_1 salt_1;
    rand keymgr_reg_salt_2 salt_2;
    rand keymgr_reg_salt_3 salt_3;
    rand keymgr_reg_salt_4 salt_4;
    rand keymgr_reg_salt_5 salt_5;
    rand keymgr_reg_salt_6 salt_6;
    rand keymgr_reg_salt_7 salt_7;
```

multi-fields will be handled in next PR

Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

**Comment by rswarbrick** [comment]
This is a nice improvement, but I don't think the approach taken in the code is something we should merge. The problem is that you add a "upward link" from a register to its parent multi-register. This kind of breaks the data model that the reggen code uses (where big things only contain smaller things).

Fortunately, it's not too hard to do this the other way around. I'm going on holiday for a week tomorrow(!). As I didn't want to leave this in limbo, I've knocked together something that I think should work better. I'll push it up as a draft PR in a second, linking back here. Please could you grab the changes there and squash them into this PR?

There are probably a couple of corners that could still be improved. In particular, I'm not happy about having to duplicate the logic for iterating over `all_regs` in the template file, but I don't know how to avoid it without passing function pointers to hunks of template, which I don't think Mako supports. It'd be great if you had a better idea than I managed!

Assuming the data model stays sensible, please don't feel you have to wait for a review from me before merging.

**Comment by rswarbrick** [comment]
Oh, one other note: this won't work if we have a "compact" multi-register that doesn't divide equally into a whole number of registers. For example, suppose there's a 16-bit value that we want 9 copies of. In that case, the last register really is of a different type because it just contains one 16-bit field, not two.

I'm perfectly happy for us to add an assertion to the MultiRegister code to disallow such things, but please could you do that before merging support for arrays? (Otherwise, I think we'll get rather confusing errors if we ever end up in this situation).

**Comment by weicaiyang** [comment]
> Oh, one other note: this won't work if we have a "compact" multi-register that doesn't divide equally into a whole number of registers. For example, suppose there's a 16-bit value that we want 9 copies of. In that case, the last register really is of a different type because it just contains one 16-bit field, not two.
> 
> I'm perfectly happy for us to add an assertion to the MultiRegister code to disallow such things, but please could you do that before merging support for arrays? (Otherwise, I think we'll get rather confusing errors if we ever end up in this situation).

thank you very much for making the PR #7191. I have squashed it into the first commit here
on the top of that, I added 2nd commit to handle unevenly divided mulitreg. Please take a look.
Here is the commit message
```
Update on the top of #7191
1. add `multireg.dv_compact` and `reg.dv_packed_fields` to support
unevenly divided fields, and some related logic
2. add `gen_dv.riname_w_idx` to get the instance name easier without
adding similiar code in multiple places
```

3rd PR is to fix all IP to use the new RAL

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl]: I don't think this is right: line 518 still uses `reg_inst`, so you'll just get some random global, which will cause a hard-to-debug error in the future (Mako's scoping rules are awful)

File: util/reggen/multi_register.py
Comment (rswarbrick) [inline util/reggen/multi_register.py:109]: This doesn't need to be a field: it's equal to `len(self.regs)`. Please change this back to a local.

File: util/reggen/register.py
Comment (rswarbrick) [inline util/reggen/register.py:184]: This is exactly the thing I'm trying to avoid: a register shouldn't need to know or care about a surrounding multi-register. Please could you try to re-jig this so that the template code that's iterating over registers and multi-registers passes this information down ("by the way, you're item 3 of a multi register called FOO") rather than storing that on the register itself.

**Review by rswarbrick — CHANGES_REQUESTED** [review]
Just reviewing the first two patches here. In the second, you've checked in an exclusion template for the crossbar: I assume that was a mistake? Also, please re-jig the data model so that the Register code doesn't know about MultiRegister. Since it's possible, I'd rather keep things composable.

File: util/reggen/multi_register.py
Comment (weicaiyang) [reply util/reggen/multi_register.py:109]: that makes sense. Removed, thanks

File: util/reggen/register.py
Comment (weicaiyang) [reply util/reggen/register.py]: it's not very convenient to get instance name. I created a function get_inst_to_reg_dict in the tpl to make it easier and also keep the register class simple. PTAL

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: I got the reg_inst in apply_regwen. But I changed it back to get the inst_name before calling apply_regwen.

File: util/reggen/multi_register.py
Comment (rswarbrick) [inline util/reggen/multi_register.py:72]: I think this should probably be `self.reg.name`, not `self.cname`. If defined, the `cname` field seems to be used for the name of the component and is the same for multiple multiregs. See `flash_ctrl.hjson`, for example.

File: util/reggen/gen_dv.py
Comment (rswarbrick) [inline util/reggen/gen_dv.py]: Please revert these two hunks: it's probably clearer to keep the reggen imports together.

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl:14]: I don't think this comment is right: I think this function returns a dictionary mapping register instance names (if that's the right phrase?) to their corresponding registers.

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl:69]: This makes sense, but I think it would be easier to understand if it stayed pure:
```python3
    if isinstance(r, MultiRegister):
      reg = r.reg
      if r.dv_compact:
        reg.fields = r.regs[0].fields
        regs = [reg]
      else:
        regs = r.regs
    else:
      regs = [r]
```
(yeah, this wouldn't fly in C/C++, but Python has weaker scoping rules and this sort of thing seems to be idiomatic there).

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl:72]: Do these lines need switching? I think we probably want a blank line between each class.

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl:96]: As above for the definition of `regs` here.

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl]: Since we've got the if/else stuff in the block above, maybe it would be clearer to move this there? That gets rid of the big ternary, which should make things a bit easier to understand.

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl]: I like this! But it can be collapsed even further now:
```
    % for r in rb.all_regs:
      % for inst, reg in get_inst_to_reg_dict(r).items():
${apply_regwen(rb, reg, inst)}\
      % endfor
    % endfor
```

File: util/reggen/uvm_reg_base.sv.tpl
Comment (rswarbrick) [inline util/reggen/uvm_reg_base.sv.tpl:559]: Again, this can be simplified a bit:
```
% for wen in rb.all_regs:
  % for wen_inst, wen_reg in get_inst_to_reg_dict(wen).items():
```

**Review by rswarbrick — CHANGES_REQUESTED** [review]

File: util/reggen/register.py
Comment (rswarbrick) [reply util/reggen/register.py]: Looks good, thanks!

File: util/reggen/gen_dv.py
Comment (weicaiyang) [reply util/reggen/gen_dv.py]: fixed

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: I used List earlier, forgot to update comments. Fixed, thanks

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: updated, thanks

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: Good catch. fixed. thanks

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: Done

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: Indeed, moved it. Thanks

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: Good idea. Fixed.

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl]: Done. Thanks

File: util/reggen/multi_register.py
Comment (weicaiyang) [reply util/reggen/multi_register.py]: Thanks for pointing this out. Fixed.

File: hw/ip/alert_handler/dv/env/alert_handler_scoreboard.sv
Comment (weicaiyang) [inline hw/ip/alert_handler/dv/env/alert_handler_scoreboard.sv:94]: @cindychip I just resolved some merging conflict in alert_handle. it's a bit strange that we only use index 2 but I didn't see index 1 is used. Hope I didn't introduce some bugs here. PTAL

**Review by rswarbrick — APPROVED** [review]

**Comment by rswarbrick** [comment]
Looks great, thanks!

File: hw/ip/alert_handler/dv/env/alert_handler_scoreboard.sv
Comment (cindychip) [reply hw/ip/alert_handler/dv/env/alert_handler_scoreboard.sv:94]: Thanks weicai for checking. LGTM

**Pull request merged**
